File: optimize-images.js
/**
 * Image Optimization Script for Hoshiya.shop
 * Converts PNG/JPG images to optimized WebP format using sharp
 * 
 * Install dependencies first: npm install sharp
 * Run: node optimize-images.js
 */

const sharp = require('sharp');
const fs = require('fs');
const path = require('path');

const IMAGES_DIR = path.join(__dirname, 'images');
const OUTPUT_DIR = path.join(IMAGES_DIR, 'webp');
const QUALITY = 85;
const EXTENSIONS = ['.png', '.jpg', '.jpeg'];

// Create output directory if it doesn't exist
if (!fs.existsSync(OUTPUT_DIR)) {
    fs.mkdirSync(OUTPUT_DIR, { recursive: true });
}

async function optimizeImage(inputPath, outputPath) {
    try {
        const stats = fs.statSync(inputPath);
        const originalSize = stats.size / 1024; // KB

        await sharp(inputPath)
            .webp({ quality: QUALITY, effort: 6 })
            .toFile(outputPath);

        const newStats = fs.statSync(outputPath);
        const newSize = newStats.size / 1024; // KB
        const savings = ((originalSize - newSize) / originalSize) * 100;

        console.log(`✓ ${path.basename(inputPath)}`);
        console.log(`  ${originalSize.toFixed(1)}KB → ${newSize.toFixed(1)}KB (${savings.toFixed(1)}% smaller)\n`);

        return { originalSize, newSize, success: true };
    } catch (error) {
        console.log(`✗ Error processing ${path.basename(inputPath)}: ${error.message}\n`);
        return { originalSize: 0, newSize: 0, success: false };
    }
}

async function main() {
    console.log('Starting image optimization...\n');

    // Get all image files
    const files = fs.readdirSync(IMAGES_DIR)
        .filter(file => {
            const ext = path.extname(file).toLowerCase();
            return EXTENSIONS.includes(ext);
        })
        .map(file => path.join(IMAGES_DIR, file));

    console.log(`Found ${files.length} images to optimize\n`);

    let successCount = 0;
    let totalOriginalSize = 0;
    let totalNewSize = 0;

    for (const inputPath of files) {
        const basename = path.basename(inputPath, path.extname(inputPath));
        const outputPath = path.join(OUTPUT_DIR, `${basename}.webp`);

        const result = await optimizeImage(inputPath, outputPath);
        
        if (result.success) {
            successCount++;
            totalOriginalSize += result.originalSize;
            totalNewSize += result.newSize;
        }
    }

    // Summary
    const totalSavings = ((totalOriginalSize - totalNewSize) / totalOriginalSize) * 100;
    
    console.log('='.repeat(60));
    console.log('Optimization Complete!');
    console.log('='.repeat(60));
    console.log(`Images processed: ${successCount}/${files.length}`);
    console.log(`Total original size: ${(totalOriginalSize / 1024).toFixed(2)} MB`);
    console.log(`Total new size: ${(totalNewSize / 1024).toFixed(2)} MB`);
    console.log(`Total savings: ${((totalOriginalSize - totalNewSize) / 1024).toFixed(2)} MB (${totalSavings.toFixed(1)}%)`);
    console.log(`\nOptimized images saved to: ${OUTPUT_DIR}`);
}

main().catch(console.error);

File: optimize_images.py
"""
Image Optimization Script for Hoshiya.shop
Converts PNG/JPG images to optimized WebP format
"""
import os
from pathlib import Path
from PIL import Image
import sys

def optimize_image(input_path, output_path, quality=85):
    """
    Convert and optimize image to WebP format
    
    Args:
        input_path: Path to source image
        output_path: Path to save WebP image
        quality: WebP quality (0-100, default 85)
    """
    try:
        with Image.open(input_path) as img:
            # Convert RGBA to RGB if necessary
            if img.mode in ('RGBA', 'LA', 'P'):
                background = Image.new('RGB', img.size, (255, 255, 255))
                if img.mode == 'P':
                    img = img.convert('RGBA')
                background.paste(img, mask=img.split()[-1] if img.mode in ('RGBA', 'LA') else None)
                img = background
            elif img.mode != 'RGB':
                img = img.convert('RGB')
            
            # Save as WebP
            img.save(output_path, 'WebP', quality=quality, method=6)
            
            # Get file sizes
            original_size = os.path.getsize(input_path) / 1024  # KB
            new_size = os.path.getsize(output_path) / 1024  # KB
            savings = ((original_size - new_size) / original_size) * 100
            
            print(f"✓ {os.path.basename(input_path)}")
            print(f"  {original_size:.1f}KB → {new_size:.1f}KB ({savings:.1f}% smaller)")
            
            return True
    except Exception as e:
        print(f"✗ Error processing {input_path}: {e}")
        return False

def main():
    # Setup paths
    images_dir = Path(__file__).parent / 'images'
    webp_dir = images_dir / 'webp'
    webp_dir.mkdir(exist_ok=True)
    
    # Image extensions to convert
    extensions = {'.png', '.jpg', '.jpeg'}
    
    # Find all images
    image_files = []
    for ext in extensions:
        image_files.extend(images_dir.glob(f'*{ext}'))
    
    print(f"Found {len(image_files)} images to optimize\n")
    
    success_count = 0
    total_original_size = 0
    total_new_size = 0
    
    for img_path in sorted(image_files):
        output_path = webp_dir / f"{img_path.stem}.webp"
        
        if optimize_image(img_path, output_path, quality=85):
            success_count += 1
            total_original_size += os.path.getsize(img_path) / 1024
            total_new_size += os.path.getsize(output_path) / 1024
        
        print()
    
    # Summary
    total_savings = ((total_original_size - total_new_size) / total_original_size) * 100
    print(f"\n{'='*60}")
    print(f"Optimization Complete!")
    print(f"{'='*60}")
    print(f"Images processed: {success_count}/{len(image_files)}")
    print(f"Total original size: {total_original_size/1024:.2f} MB")
    print(f"Total new size: {total_new_size/1024:.2f} MB")
    print(f"Total savings: {(total_original_size - total_new_size)/1024:.2f} MB ({total_savings:.1f}%)")
    print(f"\nOptimized images saved to: {webp_dir}")

if __name__ == "__main__":
    main()
